docs/architecture/grpc-for-wcf-developers/apm.md

```diff
@@ -1,11 +1,278 @@
+
+```csharp
+public class StockData : Stocks.StocksBase
+{
+    private static readonly CounterOptions GetRequestCounter = new CounterOptions
+    {
+        Name = "StockData_Get_Requests",
+        MeasurementUnit = Unit.Calls
+    };
+
+    private readonly IStockRepository _repository;
+    private readonly IMetrics _metrics;
+
+    public StockData(IStockRepository repository, IMetrics metrics)
+    {
+        _repository = repository;
+        _metrics = metrics;
+    }
+
+    public override async Task<GetResponse> Get(GetRequest request, ServerCallContext context)
+    {
+        _metrics.Measure.Counter.Increment(GetRequestCounter);
+
+        // Serve request...
+    }
+}
+```
+
+### Storing and visualizing metrics data
+
+The best way to store metrics data is in a *time-series database*, a specialized data store designed to record numerical data series marked with timestamps. The most popular of these are [Prometheus](https://prometheus.io/) and [InfluxDB](https://www.influxdata.com/products/influxdb-overview/). Microsoft Azure also provides dedicated metrics storage through the [Azure Monitor](https://docs.microsoft.com/azure/azure-monitor/overview) service.
+
+The current go-to solution for visualizing metrics data is [Grafana](https://grafana.com), which works with a wide range of storage providers including Azure Monitor, InfluxDB and Prometheus. Here is an example Grafana dashboard showing metrics from the Linkerd service mesh running the StockData sample.
+
+![Grafana dashboard](images/grafana-screenshot.png)
+
+### Metrics-based alerting
+
+The numerical nature of metrics data means that it is ideally suited to drive alerting systems, notifying developers or support engineers when a value falls outside of some defined tolerance. The platforms already mentioned all provide support for alerting via a range of options, including emails, text messages or in-dashboard visualizations.
+
+## Distributed tracing
+
+*Distributed tracing* is a relatively recent development in monitoring, which has arisen from the increasing use of microservices and distributed architectures. A single request from a client browser, application or device may be broken down into many steps and sub-requests and involve the use of many services across a network. This makes it difficult to correlate log messages and metrics with the specific request that triggered them.
+
+Although it is still a nascent technology area, distributed tracing has grown quickly in popularity and is now supported by the Cloud Native Computing Foundation, with [the Open Tracing standard](https://opentracing.io) effort attempting to provide vendor-neutral libraries for working with backends like [Jaeger](https://www.jaegertracing.io/) and [Elastic APM](https://www.elastic.co/products/apm).
```

Comment:
OpenTracing is old and busted. OpenTelemetry is the new hotness

https://thenewstack.io/opentracing-opencensus-merge-into-a-single-new-project-opentelemetry/

Reply:
I hate computers.

Reply:
Added https://github.com/dotnet/docs/pull/14438/files#diff-02130034a4b7526dbd006328cb690893R109